### ISSUE TITLE
error "Unsupported Ops of type: Dilation2D" during conversion

### ISSUE THREAD

**Issue by devorblackbeard** [tf-coreml/tf-coreml#241]
I am trying to converter to my .pb model into a .mlmodel

i got this error "Unsupported Ops of type: Dilation2D",
how can add custom layers for Dilation2D during conversion.

**Comment by aseemw** [comment]
Please see https://github.com/tf-coreml/tf-coreml/blob/master/examples/custom_layer_examples.ipynb

**Comment by devorblackbeard** [comment]
Hi thanks for the reply, Is it possible to add Dilation2D custom layers to coreml If I follow similar process as above custom_layer_examples link?

**Comment by aseemw** [comment]
Yes!

**Comment by devorblackbeard** [comment]
So I removed Dilation2d ops but now during conversion I get the following error :
Unsupported Ops of type: Pack
Is pack a supported op for coreml conversion?

**Comment by devorblackbeard** [comment]
So I have replaced the deconvolution layers with resize bilinear , n the model got converted succesfully n got loaded on the phone without any issues.

The issue is now in the discrepancy of results generated by CoreML model n by tf model file.

![index](https://user-images.githubusercontent.com/37415406/46596356-ee640580-caf9-11e8-84d7-67057528e3b3.png)
![img_d433b1ea514c-1 1](https://user-images.githubusercontent.com/37415406/46596414-15bad280-cafa-11e8-8d18-a5f438231c18.jpeg)

**Comment by devorblackbeard** [comment]
The above image is predicted by tf model n the image below is predicted by coreml.
Can you please help me explain what's the issue here??

**Comment by Raj-08** [comment]
Both images being sent to coreml and tfmodel are of same size.

**Comment by Raj-08** [comment]
This image is being sent as a part of test dataset

**Comment by aseemw** [comment]
Here are my suggestions to debug the numerical issue:

- try calling predict (in coremltools) with the argument useCPUOnly=True. That would force the network to run on CPU alone which uses higher precision. If the outputs match, it means the lower precision path on the GPU is causing some numerical inaccuracies. 

- You could convert a portion of the graph, by specifying an intermediate tensor as the output name during conversion and compare the output of TF and CoreML, which will correspond to that tensor (both will be numpy arrays). That way you can find out from which node onwards the discrepancy becomes large. 
[Netron](https://github.com/lutzroeder/netron) is a nice tool to visualize both TF and CoreML graphs and to help choose intermediate tensors for debugging.

**Comment by devorblackbeard** [comment]
Can you mention exactly where useCPUOnly options can be set to true in swift to be tested it out on phone. Also I tried training on tensorflow with float 16 and then testing in coreml but that didnt turn out to be successful either.


And there is a very strange pattern in results . As you can see here the persons rights side is always producing errors near shoulders and near the left side of the face of the person. How is this error following this trend?

![img_1409](https://user-images.githubusercontent.com/37415406/46670083-d326f200-cbee-11e8-8b6b-ba49360b44f7.jpg)
![img_1410](https://user-images.githubusercontent.com/37415406/46670089-d6ba7900-cbee-11e8-9958-c3ab285889ae.jpg)
![img_1408](https://user-images.githubusercontent.com/37415406/46670093-d9b56980-cbee-11e8-8d9e-6af8b41594d0.jpg)

**Comment by devorblackbeard** [comment]
I tried using it on a simulator as well n I had read at a few places that uses CPU by default but yet the results were the same.

**Comment by aseemw** [comment]
Do you get different results only in swift on iphone or in python on Mac as well (while using coremltools) as well? 

How are you generating the ground truth TF results (in python or in an app using tflite) ? 
There might be a difference in feeding the input image.

If you use coremltools "predict" then "useCPUOnly" argument is available.
If you are using the swift API then set the "usesCPUOnly" flag in [MLPredictionOptions](https://developer.apple.com/documentation/coreml/mlpredictionoptions)

**Comment by Raj-08** [comment]
The tf results are generated from frozen pb on GPU pc.

**Comment by Raj-08** [comment]
When i feed an image on gpu pc , i simply read an image using cv2.imread and feed it without any preprocessing. I didnt have any preprocessing in training as well. 
while converting the coreml file, I didnt pass any image scaling options as my model expects values between 0-255

**Comment by Raj-08** [comment]
I havent tried image prediction on coreml tools.  Can you please tell me what needs to  be done in case it turns out to be same (poor results)or different (better than coreml on device) .

**Comment by aseemw** [comment]
Please see some of the notebooks to see the best practices to establish the correctness of conversion to mlmodel (say this [one](https://github.com/tf-coreml/tf-coreml/blob/master/examples/style_transfer_example.ipynb))

Our goal is to make sure that given identical inputs the .pb and .mlmodel produce similar (close to identical) outputs, when both of them are run with the CPU backend. (GPU backends are not ideal for comparison initially as there might be lower precision paths). 

To be clear:

- make sure you run both .pb and .mlmodel on the same platform: i.e. on Mac with a CPU backend 
- run both models from python
- make sure you feed the identical numpy arrays  to both the .pb and .mlmodel file. The image that is read from cv can be converted to a numpy array. Initially do not use the "image_input_names" argument during tfcoreml conversion, so that the produced CoreML model will accept Multi-arrays as input. 
- the output will be a numpy array from both TF and CoreML. Compare them, say using, maximum relative error metric

(the ordering of the numpy array will be different though, since TF accepts (H,W,C) whereas CoreML accepts (C,H,W))

**Comment by aseemw** [comment]
In the .pb inference code you have:

```
input_image1 = cv2.imread('./1.jpg')
original_im = cv2.resize(input_image1,(480,480))
im=np.expand_dims(original_im,axis=0)
```
whereas in CoreML you have: 
```
image=cv2.imread('./1.jpg')
image=np.reshape(image,(3,480,480))
```

These are not even same! 
- you call cv2.resize method for TF but not for CoreML
- You don't want reshape, you want to call transpose for changing last axis channel to first axis channel. Reshape doesn't even change the data layout, it only changes the view to the underlying float pointer. You want to change the data order. 

So for CoreML, you should instead use: 
```
image=cv2.imread('./1.jpg') 
original_im = cv2.resize(image,(480,480)) # [480,480,3]
original_im = np.transpose(original_im, (2,0,1)) # [3,480,480]
```

Similarly when you get the output back from CoreML, which would be (3,480,480) you need to transpose it to shape (480,480,3) before comparing with TF.

**Comment by Raj-08** [comment]
okay so that gave me this result(using cpu on mac , .mlmodel file) , which is incorrect still and is much different than what we get in 1st image.

![222](https://user-images.githubusercontent.com/15856029/46784618-780b1180-cd4c-11e8-94f4-b8152440d538.png)

**Comment by aseemw** [comment]
Ok. 

So now instead of comparing the final tensor  `import/SemanticPredictions:0` it will be worthwhile to compare intermediate tensors, to see where the mismatch between TF and mlmodel first happens. That way the source of the error could be narrowed down to a layer.

**Comment by Raj-08** [comment]
Hi Aseem , 

We have compared the relative errors of a few layers of our network.  And i could not find a particular layer where the source of error could be made sense of. I saw the result at places where i could have visualized it . At places where it was in higher dimensions i simply subtracted them and averaged them.

Moving from bottom to up in the network :

SemanticPredictions =  -33.72565652963911 (Image )
matting/Sigmoid_1 = 0.007266705656448191 (probability scores , but despite this error being less than SemanticPredictions, it didnt mean it was visually any better , i multiplied with image in numpy and i got the same result as SemanticPredictions)
matting/Conv2D_4 = 0.1447988222206721
matting/Sigmoid 0.003414830014866212 (This part was visually the same as SemanticPredictions)
matting/logits/semantic/Conv2D = 0.0001697001579410716
matting/decoder/decoder_conv1_pointwise/Relu = 2.611582909711135e-06
matting/decoder/decoder_conv1_pointwise/Conv2D = -0.00045329757248755055
[**matting** forms my 2nd scope of network , it is more of a **regression** based network which provides values **between** **0 and 1**]


segmentation/ArgMax:0’= 0.0072048611111111115(The result till here was perfectly fine visually )
segmentation/concat_projection/Conv2D  = 1.7275435467330644e-07
[**segmentation** forms my 1st scope of network , it is **classification** network where values are **0 , 1 and 2**]

**Comment by aseemw** [comment]
Ok that's progress!! We now know that the network till 'segmentation/ArgMax:0’ tensor is fine, only after that the discrepancies begin to creep in. 
What layer type comes after that? Maybe the convertor is not properly converting that layer type or there is a mismatch in TF/CoreML's execution of that layer type. 
If you can share your network I can take a look at it.

**Comment by Raj-08** [comment]
My network is simply this:

`IMAGE >> **SEGMENTATION**MOBILENET MODEL1 >> OUTPUT1 >> MULTIPLICATION OF OUTPUT1 with 255 >> CONCAT THIS WITH IMAGE >>**MATTING**MOBILENET MODEL2 >>OUTPUT2(ERROR APPEARS HERE)>>UPSAMPLE BY BILINEAR RESIZE>>MULTIPLICATION OF OUTPUT2 WITH 255 >> CONCAT THIS WITH IMAGE>> 3 LAYERS OF 3*3 CONVOLUTIONS>> SIGMOID >> PREDICTION >> SPLIT IMAGE IN 3 CHANNELS >> MULTIPLY PREDICTION WITH ALL 3 CHANNELS >> CONCAT 3 CHANNELS >> FINAL OUTPUT`

Segmentation which forms the 1st scope (till where results are good) is the same network as Matting which forms the second scope of my network (The way they work is different , however architecture is same) , so i am not able to get how 2nd network is unable to perform whereas 1st network is performing as it should be.

The only additional operation between them is 
**multiplication** of 'segmentation/ArgMax:0’ with 255 . 
This is **concatenated** with image and passed as input to 2nd scope.

**Comment by aseemw** [comment]
Yeah it is not clear. 
Based on your observations and experiments It seems that the difference can only come from the multiplication layer. 

You can open a bug report at https://developer.apple.com/bug-reporting/ and attach the model there and share the bug number. I can then take a look at the actual model and see if I find the source of the bug.

**Comment by aseemw** [comment]
1) First make sure that you have installed tfcoreml from source (https://github.com/tf-coreml/tf-coreml#install-from-source) so that you have the latest version of the resizeBilinear layer. 

2) The problem is happening because CoreML does not have a “cast” layer and its output is in float format whereas TF’s output is in “int” format. If we look at the end portion of the TF graph we see the cast operation. So either convert till the second last tenosr “concat:0”, that matches between TF and CoreML. If you convert till the last tensor “SemanticPredictions:0”, you just have to cast the CoreML output (which is float) to int. 


![tf_model_end](https://user-images.githubusercontent.com/7546478/46978153-f9cba980-d082-11e8-925b-c36fbf6c0ba1.png)

**Comment by Raj-08** [comment]
Thank you Aseem for putting in your time , i got it to work after installing tfcoreml from source.

**Comment by Raj-08** [comment]
we will close this now

**Comment by devorblackbeard** [comment]
Thank you, I got fixed.

**Comment by sibadakesi** [comment]
> 请参阅https://github.com/tf-coreml/tf-coreml/blob/master/examples/custom_layer_examples.ipynb

The link has expired,please can you provide a new link.
